### ISSUE TITLE
Refactor FormField into separate abstract and concrete classes

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#10036]
Right now `TextFormField` extends `FormField` and that's the only place in Flutter where a StatefulWidget extends a concrete StatefulWidget. We should take the parts of `FormField` that `TextFormField` uses and refactor them into an abstract subclass of StatefulWidget that they both extend.

/cc @xster @abarth

**Comment by Hixie** [comment]
If we do this we have to be careful to do it in a backwards-compatible way to avoid breaking API stability.

**Comment by Hixie** [comment]
What's wrong with extending a concrete subclass of StatefulWidget?

**Comment by collinjackson** [comment]
- I would normally expect to be able to pass a GlobalKey<TextFormFieldState> as the `key` argument to the `TextFormField` constructor but TextFormFieldState isn't a type.

- I would normally expect that new TextFormField(autofocus: true).autofocus == true but I get autofocus is not defined.

- If we add new named constructor arguments to FormField, subclasses that had constructor arguments of the same name but with different types will result in "invalid override" errors.

- Things get more dangerous if the subclass overrides build(). If we add new named constructor arguments to FormField, subclasses that had constructor arguments of the same name and types might result in the subclass's version being intentionally used in two places; once in the subclass and once by FormField in the call to super.build().

tl;dr composition > inheritance. If we can compose, we should.

@sethladd

**Comment by Hixie** [comment]
I agree that we should fix those issues, and I'm happy to do so by not using inheritance, but I'm not convinced that those issues are implied by the use of inheritance from a concrete class.

**Comment by collinjackson** [comment]
I'm fine with this not being a hard-and-fast rule as long as there are Reasons.

**Comment by abarth** [comment]
The base class taking a `builder` is also really awkward because you need to be able to create the closure for `super(builder:...)` without access to `this`.

**Comment by Hixie** [comment]
If this would require an API change, it's not happening. But I'm leaving this open so that we can study it in more detail when we have more time.
